### PR TITLE
feat(date-picker): added date picker component

### PR DESCRIPTION
**Pull request by marijohannessen**
## Overview

Resolves https://github.ibm.com/Bluemix/design-system-website/issues/661

Added Date Picker component.

The component is using [flatpickr](https://chmln.github.io/flatpickr/) as its base for the calendar. There are 3 different variations: `no-calendar`, `basic` and `range`.

**Visual review:** http://date-picker-example.stage1.mybluemix.net/components/date-picker/

## Testing / Reviewing

Do `npm install` before reviewing.

### CODE REVIEW

File: src/components/date-picker/date-picker.js
Comment (asudoh) [inline src/components/date-picker/date-picker.js]: Different countries have different date format. I’d suggest using [`Intl.DateTimeFormat`](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/DateTimeFormat#Examples).

File: src/components/date-picker/date-picker.js
Comment (asudoh) [inline src/components/date-picker/date-picker.js]: It’s a bit of repeating, but, for example, “2017/5/31” is used in Japan instead of “5/31/2017”. Seems that Flatpickr has [custom `formatDate()` config](https://github.com/chmln/flatpickr/commit/a1c3548b7046d07c4d10e0ae8efdb01442337587) as well ass [custom `parseDate()` config](https://github.com/chmln/flatpickr/blob/a1c3548b7046d07c4d10e0ae8efdb01442337587/src/flatpickr.js#L1103), and thus I’d suggest using those - The implementation should use `Into.DateTimeFormat`.

File: package.json
Comment (hellobrian) [inline package.json]: It looks we removed this from our package.json in [`master` branch](https://github.com/carbon-design-system/carbon-components/blob/master/package.json#L35)

Right now, the current failing builds in Travis are because npm cannot find `@ibm/karma-ibma`...I've also excluded this [devDependency from my PR](https://github.com/carbon-design-system/carbon-components/pull/140) and Travis is passing. 

I'll keep digging to see if there are any other factors at play.

Comment (asudoh) [reply package.json]: @hellobrian You are right that we have removed karma-ibma dependency from package.json, as [it has to be downloaded from https://aat.mybluemix.net/dist/karma-ibma.tgz](https://github.com/carbon-design-system/carbon-components/blob/0a9da53/.travis.yml#L42) (as it’s an internal dependency).

File: src/components/date-picker/date-picker.js
Comment (marijohannessen) [reply src/components/date-picker/date-picker.js]: Hi @asudoh , I changed the `formatDate` function a bit, do you think this is good enough or do we need to add more logic to this?

**Comment by hellobrian** [comment]
A couple of NumberInput tests are failing: 

image

File: src/components/date-picker/date-picker--range-calendar.html
Comment (hellobrian) [inline src/components/date-picker/date-picker--range-calendar.html]: You can remove `xmlns` attribute, it's not needed on the browsers we support.

File: src/components/date-picker/date-picker--single-calendar.html
Comment (hellobrian) [inline src/components/date-picker/date-picker--single-calendar.html]: Even though this is for demo, make sure all associated `for` and `id` attrs are unique

**Comment by tw15egan** [comment]
Could you remove the `yarn-error.lock` file?

File: src/components/date-picker/date-picker.js
Comment (hellobrian) [inline src/components/date-picker/date-picker.js]: You shouldn't need to create a new variable for `inputValue`, you can just make the array variable:
```js
// const inputValue = inputField.value;
const inputValueArr = inputField.value.split('/');
```

And your `forEach` loop can use the `index` instead of the `increment` variable.
```js
inputValueArr.forEach((val, index) => {
  // increment++;
  if (index === inputValueArr.length) {
    newInputValue += `${val}`;
  } else {
    newInputValue += `${val} / `;
  }
});
```

**Comment by hellobrian** [comment]
@marijohannessen I think you also need to work on styles for high-contrast mode. Tested in FF and I think the range selection should be made more obvious since the background styles don't really come through.

image

File: src/components/date-picker/README.md
Comment (hellobrian) [inline src/components/date-picker/README.md]: Don't forget to update these docs with DatePicker stuffz

File: src/components/date-picker/_date-picker.scss
Comment (hellobrian) [inline src/components/date-picker/_date-picker.scss]: Make sure to test that styles work when importing `_data-picker.scss` only

Comment (marijohannessen) [reply src/components/date-picker/_date-picker.scss]: Not sure how to test this as this is the only place I'm importing `flatpickr.scss`?

File: src/components/date-picker/README.md
Comment (marijohannessen) [reply src/components/date-picker/README.md]: Will do once it's ready to be merged (in case some stuff changes) :)

**Comment by marijohannessen** [comment]
@hellobrian Addressed most of your comments! Will work on the high contrast styles, thanks for catching that. No idea why the NumberInput tests are failing, I haven't touched it :-/ Will look into that as well!

**Comment by asudoh** [comment]
Hi @marijohannessen in addition to what I mentioned earlier, I'd suggest the following changes:

1. Give the consumer of our date picker more control of Flatpickr's configuration. This is important especially for globalization perspective, given Flatpickr's configuration includes sets of localization strings. It can be done by passing along `this.options` to Flatpicker with a tweak: https://github.com/asudoh/carbon-components/commit/29b3e83ae0b7e8fda73e78df3c58ed40c7ad4088#diff-1a0dc4498ad653a61b27f80bced9f215R40
2. Separate the data attribute for date picker type from `selectorInit`, by introducing e.g. `data-date-picker-type`: https://github.com/asudoh/carbon-components/commit/29b3e83ae0b7e8fda73e78df3c58ed40c7ad4088#diff-1a0dc4498ad653a61b27f80bced9f215R28 https://github.com/asudoh/carbon-components/commit/29b3e83ae0b7e8fda73e78df3c58ed40c7ad4088#diff-1a0dc4498ad653a61b27f80bced9f215R206
  This is for allowing inherited classes to use different `selectorInit` - Otherwise only one or the other (parent class or inherited class) can be used in an application.
3. Destroying Flatpickr upon releasing our date picker: https://github.com/asudoh/carbon-components/commit/8686951a0b279f953b77929210453d944a3c4fd0#diff-1a0dc4498ad653a61b27f80bced9f215R178
  BTW I don't think the reference of Flatpickr instance has to be attached to DOM element, because application can grab our date picker instance any time via `CarbonComponents.DatePicker.components.get(element)`.

Another thing I found was that the dropdown (Flatpickr) does not seem to reflect what user directly typed in the input box - Though I haven't had time to deep give in there.

**Comment by asudoh** [comment]
^ Another thing I'd suggest is going away from what `._addInputLogic()` does - It appears that it's for mangling date format (my apologies if I'm wrong here), but as I mentioned earlier - Different countries have different date format and the date format mangling logic looks very specific to the date format in the united states.

File: src/globals/scss/styles.scss
Comment (asudoh) [inline src/globals/scss/styles.scss]: @marijohannessen Did you really want this line? Seems that this is for time picker and I'm getting a build error complaining that this SCSS file is missing.

**Comment by asudoh** [comment]
Re: https://github.com/carbon-design-system/carbon-components/pull/143#issuecomment-305717466

> Another thing I found was that the dropdown (Flatpickr) does not seem to reflect what user directly typed in the input box

I found that user has to hit Enter to get the date picker updated upon input value: https://github.com/chmln/flatpickr/issues/828#issuecomment-302886760
As the issue suggests, I'd recommend hooking the `change` event to update the date picker.

File: src/globals/scss/styles.scss
Comment (marijohannessen) [reply src/globals/scss/styles.scss]: Yeah I noticed this as well, will remove it! Thanks!

**Comment by marijohannessen** [comment]
Thanks for your comments @asudoh , will work through them today and let you know when the PR is ready for another review.

**Comment by marijohannessen** [comment]
@asudoh just updated the PR with your suggestions, let me know if it looks okay!

File: src/components/date-picker/date-picker.js
Comment (asudoh) [inline src/components/date-picker/date-picker.js]: The `.release()` code refers to `this.calendar` whereas the code here refers to `this.element.calendar`. Given `DatePicker` instance associated with a DOM element can be referred to via `DatePicker.components.get(element)`, I don't think we have to attach Flatpickr instance to DOM element. Also given this I guess some of the getters/setters for interacting with Flatpickr instance (especially ones doing direct mappings) is not necessary?

File: src/components/date-picker/date-picker.js
Comment (asudoh) [inline src/components/date-picker/date-picker.js]: Given we now let Flatpickr to format/parse dates - Do we still need the logic here? Note that we can use `String.prototype.trim()` for removing preceding/trailing spaces.

File: src/components/date-picker/date-picker.js
Comment (asudoh) [inline src/components/date-picker/date-picker.js]: Letting `Date` constructor to parse date will cause globalization issues. I'd suggest using `parseDate(string, format)` API of Flatpickr to align to the parse logic Flatpickr runs internally.

**Review by asudoh — COMMENTED** [review]
Thanks @marijohannessen for making updates! Added some more comments for code consistency and globalization.

File: src/components/date-picker/_date-picker.scss
Comment (hellobrian) [reply src/components/date-picker/_date-picker.scss]: Sorry, I meant for `_date-picker.scss`. Make sure that this has all of the dependencies it needs so that users can do `@import 'date-picker'` only.

You can do test this by doing the following:
- comment out all the `@import` statements in 
- comment out the extra styles in `demo.scss`

File: src/components/date-picker/date-picker.js
Comment (marijohannessen) [reply src/components/date-picker/date-picker.js]: Ah good point! I'll remove the DOM element association, and some of the getters/setters. Thanks!

File: src/components/date-picker/date-picker.js
Comment (marijohannessen) [reply src/components/date-picker/date-picker.js]: I don't think we do! It seems to be working fine without it, so I removed it. Thanks!

File: src/components/date-picker/date-picker.js
Comment (marijohannessen) [reply src/components/date-picker/date-picker.js]: Okay, I updated it, let me know if this is what you were thinking!

**Comment by marijohannessen** [comment]
Updated the PR again @asudoh , let me know if you see any more issues!

File: src/components/date-picker/_date-picker.scss
Comment (marijohannessen) [reply src/components/date-picker/_date-picker.scss]: Ah okay, fixed it now! Thanks :)

File: src/components/date-picker/date-picker.js
Comment (asudoh) [inline src/components/date-picker/date-picker.js]: Rather than checking the length which is format-specific, I’d suggest running `setDate()` *only* when `parseDate()` is successful.

**Review by asudoh — COMMENTED** [review]
Thanks alot @marijohannessen for making the changes! The code looks good now except one I put during this round of review.

File: src/components/date-picker/date-picker.js
Comment (marijohannessen) [reply src/components/date-picker/date-picker.js]: Ah good idea, will fix! Thank you!

File: src/components/date-picker/date-picker.js
Comment (asudoh) [inline src/components/date-picker/date-picker.js]: @marijohannessen My apologies that not catching this earlier, but the logic trying to pick up the first character will be a problem for non-English users. The code here assumes that the days of week with 1st letter ambiguity is Tuesday and Thursday, but it'd be true only in English - For example, days of week in Chinese is "星期一", "星期二", ... (I guess they mean "first day of the week", "second day of the week", ...) where the 1st characters of them are the same for all days of week.

One option I can think of here is overriding `Flatpickr.l10n.en.weekdays.shorthand` outside `DatePicker` class definition instead.

**Review by asudoh — COMMENTED** [review]
Hi @marijohannessen thank you for making another series of changes! Found just one thing, the rest of the JavaScript code looks good. Thanks!

File: src/components/date-picker/date-picker.js
Comment (marijohannessen) [reply src/components/date-picker/date-picker.js]: Ah okay, I didn't think of that! Thanks. Will fix!

File: gulpfile.js
Comment (asudoh) [inline gulpfile.js]: Need to sort out the a11y issues the team hit. Also, if you simply disable a11y test in Travis, you can change .travis.yml instead of Gulp general test task.

**Review by asudoh — CHANGES_REQUESTED** [review]
The JavaScript code looks good to go once we revert general Gulp test task. Thank you very much for all the efforts to get data picker in!

**Comment by marijohannessen** [comment]
screen shot 2017-06-09 at 11 25 45 am

File: src/components/date-picker/README.md
Comment (tw15egan) [inline src/components/date-picker/README.md]: Is there supposed to be something inside the ``? Also, I think it should be phrased `come with Flatpickr are`

File: src/components/date-picker/_date-picker.scss
Comment (tw15egan) [inline src/components/date-picker/_date-picker.scss:49]: Since this seems to be re-used on all input elements, this might have enough justification to create a mixin to set the invisible borders. Also makes theming easier, as it can be set globally.

File: src/components/date-picker/date-picker--range.html
Comment (tw15egan) [inline src/components/date-picker/date-picker--range.html]: For consistency, should this be `__container` or `-container`? I thought `__` was only used when we are positioning an element relatively? Correct me if I'm wrong

File: src/components/form/_form.scss
Comment (tw15egan) [inline src/components/form/_form.scss]: Not sure if this solution will work, as this will cause inconsistent heights when using multiple form items. The text will have a height, even when hidden using this method.

File: src/components/form/_form.scss
Comment (tw15egan) [inline src/components/form/_form.scss]: Same ^

**Review by tw15egan — COMMENTED** [review]
👍 ✅ ❗️

File: src/components/date-picker/README.md
Comment (marijohannessen) [reply src/components/date-picker/README.md]: Ah yeah, I'll update the docs!

File: src/components/date-picker/_date-picker.scss
Comment (marijohannessen) [reply src/components/date-picker/_date-picker.scss:49]: Good idea! Should we do that in a separate PR maybe?

File: src/components/date-picker/date-picker--range.html
Comment (marijohannessen) [reply src/components/date-picker/date-picker--range.html]: I think it's: `main-element__element` and `main-element--modifier` ? And then in the SCSS you just don't nest the BEM elements.

File: src/components/form/_form.scss
Comment (marijohannessen) [reply src/components/form/_form.scss]: Yeah I didn't think about that. Will fix!

File: src/components/date-picker/_date-picker.scss
Comment (tw15egan) [reply src/components/date-picker/_date-picker.scss:49]: Yeah, we can just make an issue for now

File: src/components/date-picker/date-picker--range.html
Comment (hellobrian) [reply src/components/date-picker/date-picker--range.html]: FWIW, I thought we were doing it the way @tw15egan said. 
cc: @chrisdhanaraj @iangfleming

**Comment by hellobrian** [comment]
Merge conflict

File: src/components/date-picker/README.md
Comment (hellobrian) [inline src/components/date-picker/README.md]: Can you also include the date format that the regex pattern expect? 

For example, docs can be edited to say:
The default regex pattern that ships with
simple date picker is `\d{1,2}/\d{4}` ('dd/yyyy' for short date pickers) and `\d{1,2}/\d{1,2}/\d{4}` ('dd/mm/yyyy' or mm/dd/yyyy)

**Review by hellobrian — APPROVED** [review]
Small change to docs, then 👍 ✅ ❗️

File: demo/views/demo-all.dust
Comment (asudoh) [inline demo/views/demo-all.dust]: @marijohannessen I quickly ran `npm run dev` using demo.js instead of carbon-components.min.js and it seems to be working. Do you still have the issues here?

File: demo/views/demo-all.dust
Comment (asudoh) [inline demo/views/demo-all.dust]: Seems that this line is for [L51](https://github.com/carbon-design-system/carbon-components/pull/143/files#diff-627e32a3f2084e6bc734d3d244058336R51)-[L53](https://github.com/carbon-design-system/carbon-components/pull/143/files#diff-627e32a3f2084e6bc734d3d244058336R53) explanations, so we have two options here:

1. Removing [L51](https://github.com/carbon-design-system/carbon-components/pull/143/files#diff-627e32a3f2084e6bc734d3d244058336R51)-[L53](https://github.com/carbon-design-system/carbon-components/pull/143/files#diff-627e32a3f2084e6bc734d3d244058336R53) altogether
2. Keeping this line

File: demo/views/demo-all.dust
Comment (hellobrian) [reply demo/views/demo-all.dust]: Using `<script src="/carbon-components.min.js"></script>` works for me. I didn't need to add `CarbonComponents.settings.disableAutoInit = false;` either.

Comment (marijohannessen) [reply demo/views/demo-all.dust]: It works when I run this locally, but it fails when I deploy it to Bluemix for visual review.. So we need to think of a process where it will work both ways.

File: demo/views/demo-all.dust
Comment (marijohannessen) [reply demo/views/demo-all.dust]: I ended up just removing those lines, since Brian said that it worked without having to disable the auto init.

**Pull request merged**
